### PR TITLE
migrating feature field  to array field 

### LINKED ISSUE #2215: Convert `Facility.features` to ArrayField
the `features` field is a `MultiSelectField` which uses `CharField` to store data in db, convert the data to a list of integers and store as `ArrayField[SmallIntegerField]`, 
create necessary migrations.

### PR DESCRIPTION
**Pull request by praffq**
## Proposed Changes

Migrations sequence -
- renamed feature to old_feature
- added a feature field with array field
- copied data from old_feature field to feature field
- deleted old_feature field

### Associated Issue

- fixes #2215 

## Merge Checklist

- [ ] Tests added/fixed
- [ ] Update docs in `/docs`
- [ ] Linting Complete
- [ ] Any other necessary step

_*Only PR's with test cases included and passing lint and test pipelines will be reviewed*_

@coronasafe/care-backend-maintainers @coronasafe/care-backend-admins

### CODE REVIEW

File: care/facility/api/serializers/facility.py [hunk 1]
```diff
@@ -95,7 +98,10 @@ class FacilitySerializer(FacilityBasicInfoSerializer):
     # }
     read_cover_image_url = serializers.URLField(read_only=True)
     # location = PointField(required=False)
-    features = serializers.MultipleChoiceField(choices=FEATURE_CHOICES)
+    features = serializers.ListField(
+        child=serializers.ChoiceField(choices=FEATURE_CHOICES),
+        required=False,
+    )
```
Comment (sainak) [inline care/facility/api/serializers/facility.py:105]: can you check if the validations are working 
- should not allow values not present in conices
- should not allow duplicates

File: care/facility/migrations/0438_rename_features_facility_old_features.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+from django.db import migrations, models
+
+
+def convert_features_to_array(apps, schema_editor):
+    Facility = apps.get_model("facility", "Facility")
+
+    for facility in Facility.objects.all():
+        if facility.old_features:
+            facility.features = list(map(int, facility.old_features))
+            facility.save()
```
Comment (sainak) [inline care/facility/migrations/0438_rename_features_facility_old_features.py]: use bulk update

File: care/facility/migrations/0438_rename_features_facility_old_features.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+                blank=True,
+                null=True,
+                size=None,
+            ),
+        ),
+        migrations.RunPython(convert_features_to_array),
+        migrations.RemoveField(
+            model_name="facility",
+            name="old_features",
+        ),
```
Comment (sainak) [inline care/facility/migrations/0438_rename_features_facility_old_features.py]: we can remove the field in a later pr

File: care/facility/models/facility.py [hunk 4]
```diff
@@ -160,11 +159,10 @@ class Facility(FacilityBaseModel, FacilityPermissionMixin):
...
     kasp_empanelled = models.BooleanField(default=False, blank=False, null=False)
-    features = MultiSelectField(
-        choices=FEATURE_CHOICES,
-        null=True,
+    features = ArrayField(
+        models.SmallIntegerField(choices=FEATURE_CHOICES),
         blank=True,
-        max_length=get_max_length(FEATURE_CHOICES, None),
+        null=True,
     )
```
Comment (sainak) [inline care/facility/models/facility.py]: add the get_features_display method as the array field doesn't register this getter

**Review by sainak — CHANGES_REQUESTED** [review]
we can remove old_features in a later pr

File: care/facility/models/facility.py (hunk 4 above)
Comment (praffq) [reply care/facility/models/facility.py]:
```
    @property
    def get_features_display(self):
        if not self.features:
            return []
        return [REVERSE_FEATURE_CHOICES[f] for f in self.features]
```

something like this ?

Comment (sainak) [reply care/facility/models/facility.py]: yes

**Comment by praffq** [comment]
@sainak  made all the required changes, please take a look

File: care/facility/models/facility.py [hunk 5]
```diff
@@ -50,6 +51,7 @@
 ROOM_TYPES.extend(BASE_ROOM_TYPES)
 
 REVERSE_ROOM_TYPES = reverse_choices(ROOM_TYPES)
+REVERSE_FEATURE_CHOICES = reverse_choices(FEATURE_CHOICES)
```
Comment (sainak) [inline care/facility/models/facility.py:65]: we can use a new integer choice enum `FacilityFeature`
keep the old one in place, just replace it for the new field

File: care/facility/migrations/0438_rename_features_facility_old_features.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+def convert_features_to_array(apps, schema_editor):
+    Facility = apps.get_model("facility", "Facility")
+
+    facilities_to_update = Facility.objects.filter(old_features__isnull=False)
+
+    updated_facilities = []
+
+    for facility in facilities_to_update:
+        facility.features = list(map(int, facility.old_features.split(",")))
+        updated_facilities.append(facility)
```
Comment (sainak) [inline care/facility/migrations/0438_rename_features_facility_old_features.py]:
```suggestion
        try:
            facility.features = list(map(int, facility.old_features.split(",")))
            updated_facilities.append(facility)
        except ValueError:
            print(f"facility '{facility.name)' has invalid facility features")
```

File: care/facility/migrations/0438_rename_features_facility_old_features.py
Comment (sainak) [inline care/facility/migrations/0438_rename_features_facility_old_features.py:1]: update the migration number and remove the merge migrations

**Review by rithviknishad — APPROVED** [review]

**Comment by praffq** [comment]
@sainak  @rithviknishad  made the changes , can you please run the workflows to check if everything's alright?

File: care/facility/migrations/0445_rename_features_facility_old_features.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        except ValueError:
+            print(f"facility '{facility.name}' has invalid facility features")
+
+    if updated_facilities:
+        Facility.objects.bulk_update(updated_facilities, ["features"])
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("facility", "0437_alter_dailyround_rounds_type"),
```
Comment (rithviknishad) [inline care/facility/migrations/0445_rename_features_facility_old_features.py]: this needs to be changed to the previous migration before it


```suggestion
        ("facility", "0444....."),
```

File: care/facility/migrations/0446_merge_20240712_1357.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ("facility", "0444_alter_medicineadministration_dosage_and_more"),
+        ("facility", "0445_rename_features_facility_old_features"),
+    ]
+
+    operations = []
```
Comment (rithviknishad) [inline care/facility/migrations/0446_merge_20240712_1357.py]: That way this merge migration can be deleted.

**Review by rithviknishad — CHANGES_REQUESTED** [review]

**Comment by praffq** [comment]
updated. Should be fine now.

**Comment by praffq** [comment]
@rithviknishad @sainak  I've fixed the errors , updated the migration file name and dependencies , also fixed the static dummy data from facility.json

**Review by rithviknishad — APPROVED** [review]

**Review by sainak — APPROVED** [review]

**Review by vigneshhari — APPROVED** [review]

**Pull request merged**